crates/engine/tree/src/download.rs

```diff
@@ -29,6 +46,241 @@ pub enum DownloadOutcome {
     Blocks(Vec<SealedBlockWithSenders>),
 }
 
+/// A wrapper type around [`SealedBlockWithSenders`] that implements the [Ord]
+/// trait by block number.
+#[derive(Debug, Clone, PartialEq, Eq)]
+struct OrderedSealedBlockWithSenders(SealedBlockWithSenders);
+
+impl PartialOrd for OrderedSealedBlockWithSenders {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for OrderedSealedBlockWithSenders {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.0.number.cmp(&other.0.number)
+    }
+}
+
+impl From<SealedBlock> for OrderedSealedBlockWithSenders {
+    fn from(block: SealedBlock) -> Self {
+        let senders = block.senders().unwrap_or_default();
+        Self(SealedBlockWithSenders { block, senders })
+    }
+}
+
+impl From<OrderedSealedBlockWithSenders> for SealedBlockWithSenders {
+    fn from(value: OrderedSealedBlockWithSenders) -> Self {
+        let senders = value.0.senders;
+        Self { block: value.0.block, senders }
+    }
+}
+
+/// Basic [BlockDownloader].
+pub struct BasicBlockDownloader<Client>
+where
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    /// A downloader that can download full blocks from the network.
+    full_block_client: FullBlockClient<Client>,
+    /// In-flight full block requests in progress.
+    inflight_full_block_requests: Vec<FetchFullBlockFuture<Client>>,
+    /// In-flight full block _range_ requests in progress.
+    inflight_block_range_requests: Vec<FetchFullBlockRangeFuture<Client>>,
+    /// Sender for engine events.
+    event_sender: EventSender<BeaconConsensusEngineEvent>,
+    /// Buffered blocks from downloads - this is a min-heap of blocks, using the block number for
+    /// ordering. This means the blocks will be popped from the heap with ascending block numbers.
+    set_buffered_blocks: BinaryHeap<Reverse<OrderedSealedBlockWithSenders>>,
+    /// Engine download metrics.
+    metrics: BlockDownloaderMetrics,
+}
+
+impl<Client> BasicBlockDownloader<Client>
+where
+    Client: HeadersClient + BodiesClient + Clone + Unpin + 'static,
+{
+    /// Create a new instance
+    pub(crate) fn new(
+        client: Client,
+        chain_spec: Arc<ChainSpec>,
```

Comment:
ok, changed here https://github.com/paradigmxyz/reth/pull/9118/commits/36730cf14debcffc13dc0b1eede85b9b82a118ee to use consensus as constructor arg instead of chainspec, ptal